File: relay/src/main/kotlin/com/diabdata/relay/models/Messages.kt
package com.diabdata.relay.models

import kotlinx.serialization.Serializable

// ═══════════════════════════════════════════
//  CONSTANTES DE TYPES
// ═══════════════════════════════════════════

object MessageType {
    const val REGISTER = "REGISTER"
    const val REGISTER_OK = "REGISTER_OK"
    const val UNREGISTER = "UNREGISTER"
    const val AUTH = "AUTH"
    const val AUTH_OK = "AUTH_OK"
    const val AUTH_FAILED = "AUTH_FAILED"
    const val REQUEST = "REQUEST"
    const val FORWARD = "FORWARD"
    const val RESPONSE = "RESPONSE"
    const val SESSION_CLOSED = "SESSION_CLOSED"
    const val ERROR = "ERROR"
}

// ═══════════════════════════════════════════
//  ANDROID APP → RELAI
// ═══════════════════════════════════════════

@Serializable
data class RegisterMessage(
    val type: String = MessageType.REGISTER,
    val sessionId: String,
    val tokenHash: String,
    val mode: String,
    val expiresAt: Long
)

@Serializable
data class UnregisterMessage(
    val type: String = MessageType.UNREGISTER,
    val sessionId: String
)

@Serializable
data class AppResponseMessage(
    val type: String = MessageType.RESPONSE,
    val requestId: String,
    val clientId: String,
    val payload: String
)

// ═══════════════════════════════════════════
//  FRONT → RELAI
// ═══════════════════════════════════════════

@Serializable
data class AuthMessage(
    val type: String = MessageType.AUTH,
    val tokenHash: String
)

@Serializable
data class RequestMessage(
    val type: String = MessageType.REQUEST,
    val requestId: String,
    val payload: String
)

// ═══════════════════════════════════════════
//  RELAI → ANDROID APP
// ═══════════════════════════════════════════

@Serializable
data class RegisterOkMessage(
    val type: String = MessageType.REGISTER_OK,
    val sessionId: String
)

@Serializable
data class ForwardMessage(
    val type: String = MessageType.FORWARD,
    val requestId: String,
    val clientId: String,
    val payload: String
)

// ═══════════════════════════════════════════
//  RELAI → FRONT
// ═══════════════════════════════════════════

@Serializable
data class AuthOkMessage(
    val type: String = MessageType.AUTH_OK,
    val mode: String
)

@Serializable
data class AuthFailedMessage(
    val type: String = MessageType.AUTH_FAILED,
    val reason: String
)

@Serializable
data class ClientResponseMessage(
    val type: String = MessageType.RESPONSE,
    val requestId: String,
    val payload: String
)

@Serializable
data class SessionClosedMessage(
    val type: String = MessageType.SESSION_CLOSED,
    val reason: String
)

// ═══════════════════════════════════════════
//  UTILITAIRE
// ═══════════════════════════════════════════

@Serializable
data class BaseMessage(
    val type: String
)

File: src/types/messages.ts
export const MessageType = {
    REGISTER: "REGISTER",
    REGISTER_OK: "REGISTER_OK",
    UNREGISTER: "UNREGISTER",
    UNREGISTER_OK: "UNREGISTER_OK",
    AUTH: "AUTH",
    AUTH_OK: "AUTH_OK",
    AUTH_FAILED: "AUTH_FAILED",
    REQUEST: "REQUEST",
    FORWARD: "FORWARD",
    RESPONSE: "RESPONSE",
    SESSION_CLOSED: "SESSION_CLOSED",
    ERROR: "ERROR"
} as const

export type MessageTypes = typeof MessageType[keyof typeof MessageType]

export const SessionInterruptionType = {
    USER_ENDED: "USER_ENDED",
    SESSION_EXPIRED: "SESSION_EXPIRED",
    APP_DISCONNECTED: "APP_DISCONNECTED"
} as const

export type SessionInterruptionTypes = typeof SessionInterruptionType[keyof typeof SessionInterruptionType]

export const SharingType = {
    COMPANION: "COMPANION",
    MEDICAL: "MEDICAL"
} as const

export type SharingMode = typeof SharingType[keyof typeof SharingType]

export const ErrorType = {
    SESSION_ALREADY_EXISTS: "SESSION_ALREADY_EXISTS",
    SESSION_DOES_NOT_EXIST: "SESSION_DOES_NOT_EXIST",
    SESSION_COULD_NOT_BE_TERMINATED: "SESSION_COULD_NOT_BE_TERMINATED",
    SESSION_COULD_NOT_BE_CREATED: "SESSION_COULD_NOT_BE_CREATED",
    MISSING_TOKEN_HASH: "MISSING_TOKEN_HASH"
} as const

export type ErrorTypes = typeof ErrorType[keyof typeof ErrorType]

// ═══════════════════════════════════════════
//  ANDROID APP → RELAY
// ═══════════════════════════════════════════

export interface RegisterMessage {
    type: typeof MessageType.REGISTER;
    sessionId: string;
    tokenHash: string;
    mode: SharingMode;
    expiresAt: number
}

export interface UnregisterMessage {
    type: typeof MessageType.UNREGISTER;
    sessionId: string;
}

export interface AppResponseMessage {
    type: typeof MessageType.RESPONSE;
    requestId: string;
    clientId: string;
    payload: string;
}

// ═══════════════════════════════════════════
//  FRONT → RELAY
// ═══════════════════════════════════════════

export interface AuthMessage {
    type: typeof MessageType.AUTH;
    tokenHash: string;
}

export interface RequestMessage {
    type: typeof MessageType.REQUEST;
    requestId: string;
    payload: string;
}

// ═══════════════════════════════════════════
//  RELAY → ANDROID APP
// ═══════════════════════════════════════════

export interface RegisterOkMessage {
    type: typeof MessageType.REGISTER_OK;
    sessionId: string;
}

export interface ForwardMessage {
    type: typeof MessageType.FORWARD;
    requestId: string;
    clientId: string;
    payload: string;
}

// ═══════════════════════════════════════════
//  RELAI → FRONT
// ═══════════════════════════════════════════

export interface AuthOkMessage {
    type: typeof MessageType.AUTH_OK;
    mode: SharingMode;
}

export interface AuthFailedMessage {
    type: typeof MessageType.AUTH_FAILED;
    reason: string;
}

export interface ClientResponseMessage {
    type: typeof MessageType.RESPONSE;
    requestId: string;
    payload: string;
}

export interface SessionClosedMessage {
    type: typeof MessageType.SESSION_CLOSED;
    reason: string;
}

// ═══════════════════════════════════════════
//  UTILITAIRE
// ═══════════════════════════════════════════

export interface BaseMessage {
    type: MessageTypes;
}

export interface ErrorMessage {
    type: typeof MessageType.ERROR;
    reason: string;
}